cpp/src/arrow/compute/exec/spilling_util.cc

```diff
@@ -0,0 +1,253 @@
+        return Status::IOError();
+
+    HANDLE file_handle = CreateFileA(
+        tmp_name_buf,
+        GENERIC_READ | GENERIC_WRITE | FILE_APPEND_DATA,
+        0,
+        NULL,
+        CREATE_ALWAYS,
+        FILE_FLAG_NO_BUFFERING | FILE_FLAG_OVERLAPPED | FILE_FLAG_DELETE_ON_CLOSE,
+        NULL);
+    if(file_handle == INVALID_HANDLE_VALUE)
+        return Status::IOError("Failed to create temp file");
+    return file_handle;
+}
+
+static Status CloseTemporaryFile(FileHandle handle)
+{
+    if(!CloseHandle(handle))
+        return Status::IOError("Failed to close temp file");
+    return Status::OK();
+}
+
+static Status WriteBatch_PlatformSpecific(FileHandle handle, const SpillFile::BatchInfo &info)
+{
+    OVERLAPPED overlapped;
+    int64_t offset = info.start;
+    for(const ArrayInfo &arr : info.arrays)
+    {
+        for(int i = 0; i < 3; i++)
+        {
+            if(info.bufs[i] != 0)
+            {
+                overlapped.Offset = static_cast<DWORD>(offset);
+                overlapped.OffsetHigh = static_cast<DWORD>(offset >> 32);
+                if(!WriteFile(
+                       handle,
+                       info.bufs[i]->data(),
+                       info.bufs[i]->size(),
+                       NULL,
+                       &overlapped))
+                    return Status::IOError("Failed to spill!");
+                offset += info.sizes[i];
+            }
+        }
+    }
+    return Status::OK();
+}
+#else
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+Result<FileHandle> OpenTemporaryFile()
+{
+    const char *selectors[] = { "TMPDIR", "TMP", "TEMP", "TEMPDIR" };
+    constexpr size_t kNumSelectors = sizeof(selectors) / sizeof(selectors[0]);
+#ifdef __ANDROID__
+    const char *backup = "/data/local/tmp/";
+#else
+    const char *backup = "/tmp/";
+#endif
+    const char *tmp_dir = backup;
+    for(size_t i = 0; i < kNumSelectors; i++)
+    {
+        const char *env = getenv(selectors[i]);
+        if(env)
+        {
+            tmp_dir = env;
+            break;
+        }
+    }
+    size_t tmp_dir_length = std::strlen(tmp_dir);
+
+    const char *tmp_name_template = "/ARROW_TMP_XXXXXX";
+    size_t tmp_name_length = std::strlen(tmp_name_template);
+
+    constexpr int kFileNameSize = 1024;
+
+    if((tmp_dir_length + tmp_name_length) >= kFileNameSize)
+    {
+        tmp_dir = backup;
+        tmp_dir_length = std::strlen(backup);
+    }
+
+    char name[kFileNameSize + 1] = {};
+    std::strncpy(name, tmp_dir, kFileNameSize);
+    std::strncpy(name + tmp_dir_length, tmp_name_template, kFileNameSize - tmp_dir_length);
+
+#ifdef __APPLE__
+    int fd = mkstemp(name);
+    if(fd == kInvalidHandle)
+        return Status::IOError(strerror(errno));
+    if(fcntl(fd, F_NOCACHE, 1) == -1)
+        return Status::IOError(strerror(errno));
+#else    
+    int fd = mkostemp(name, O_DIRECT);
```

Comment:
There was some discussion of direct I/O in #13640.  I think there are concerns and it should at least be opt-in until we have a better picture of real world performance.  Especially considering the alignment requirements.

Reply:
Alignment requirements are not an issue since we're partitioning everything anyway. We will just make the partitioned batches aligned. 

I didn't understand that discussion, the point of spilling is to get the stuff out of memory as soon as possible and avoid memcpy-ing it around into staging buffers.

Reply:
> the point of spilling is to get the stuff out of memory as soon as possible and avoid memcpy-ing it around into staging buffers.

This is good point.

> Alignment requirements are not an issue since we're partitioning everything anyway. We will just make the partitioned batches aligned.

Sounds good.